torchbox/wagtail-torchbox
# Migration of content away from existing SubServicePages

[Link to Ticket](https://torchbox.monday.com/boards/1192293412/pulses/1242063350)

### Description of Changes Made

**TLDR;**

This PR builds up on work done in #516 by adding some tooling to help us migrate content away from SubServicePage models to SubPropositionPage models.

**Rationale**

The idea here is to automate the process of migrating content as much as possible, while leaving room for manual checking and cleanup afterwards (which should be minimal, as all the heavylifting should already have been done).

> **Note**
>
> Running this on production should ideally be done when the site is not very busy — it might be safer to temporarily put the site in maintenance mode.

**Details**

There are two management commands:

1. `./manage.py migrate_subservicepages` → running this does the following:
    - check if a migration has already been run. If it has, do nothing, otherwise, proceed.
    - for each SubServicePage:
      - copy all the content so that we can use it to create a new SubPropositionPage
      - add a suffix to the title and slug, then unpublish the page if it was live
      - create a corresponding SubPropositionPage at the same location, with all the content copied earlier
    - create a record of the migration in the database, which is used to determine whether a migration has already been run.
    - check all content for references to the SubServicePages, and replace these references with the new SubPropositionPages. This is done in
      - StreamFields
      - RichTextFields
      - ~~ForeignKeys to `"wagtailcore.Page"` or `"services.SubServicePage"`~~ not done due to time constraints
      - ~~Settings models (Site settings, Navigation settings)~~ not necessary as SubServicePages aren't referenced here
    - optionally send CSV reports via email to any specified recipients. This reports documents changes made, and areas which need manual intervention.
2. `./manage.py undo_subservicepage_migration` → this reverses the changes made in 1. above.

> **Warning**
>
> 1. Any draft changes on SubServicePages will not be transferred to the new SubPropositionPage. This will have to be done manually.
> 2. All revisions will not be transferred to the new page.
> 3. For the undo operation, we have no control on changes made to pages after the migration.

**Additional info**

> **Note**
>
> - I added a commit (372427f1) to replace `propositions.TestimonialBlock` with `core.TestimonialBlock`, because of changes made in 6518fb44, #516   
> - I added a commit (7f19575) to fix WorkPage `date` field. 

### How to Test

1. Create a backup of the database (locally, you could use [DSLR](https://github.com/mixxorz/DSLR))
2. If running locally, ensure you have staging data. Go to Wagtail Admin and take note of the SubServicePages
3. Then

    ````bash
    # omit `--send-to ...` if you don't wanna email a report
    # note that you can send report to multiple recipients by adding more
    # email addresses separated by a space 
    ./manage.py migrate_subservicepages --send-to name@example.com
    ```

4. Go to wagtail admin and check that 

    - all the SubServicePages are now unpublished, and have a suffix added to their titles and slugs
    - for each unpublished SubServicePage, there's a corresponding SubPropositionPage at the same location

5. View the new SubPropositionPages and compare them with the SubServicePages -- they should be identical
6. Take note of the pages which require manual checking of links to avert 404s (there should only be 2 pages with the following titles
    - "An Extra 3 Million Organic Clicks a Month for the NHS Website"
    - "Raising millions for Islamic Relief UK during Ramadan"
8. Need to find a way to check for 404s on the entire site

To undo migrations

```bash
manage.py undo_subservicepage_migration
```

For easier testing of emails locally, you could use [maildev](https://github.com/maildev/maildev)

Add the following to `docker-compose.yaml`

```yaml
  maildev:
    image: maildev/maildev
    ports:
      - "1080:1080"
```

And update `settings/local.py` as follows

```py
EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"

# https://docs.djangoproject.com/en/4.2/ref/settings/#email-host
EMAIL_HOST = "maildev"
# https://docs.djangoproject.com/en/4.2/ref/settings/#email-port
EMAIL_PORT = 1025
```

### Screenshots

<details>
  <summary>./manage.py migrate_subservicepages</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/96236f15-2588-42c9-a174-6a8cb57d9b66)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/755c0e60-7a0f-4f67-83b1-8005f3b8d3ee)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/e9dd79b3-bfd1-46c1-a2da-6ffa674d573d)

</details>

<details>
  <summary>./manage.py undo_subservicepage_migration</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/da4352c7-ce1f-40be-a6aa-2ac3bd1bfa2e)

</details>

<details>
  <summary>Wagtail Admin</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/f7d6f1ff-8753-4fa8-956d-ec526554bf83)

</details>

<details>
  <summary>Page comparison</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/bb225f35-1653-4173-b659-627ab58c5e58)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/bee7f6b7-e5d0-4001-854b-076069e62fb6)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/6aad2590-9956-42a7-8b55-d2b929a39207)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/de425e3a-f805-46cd-9ba0-e27f9184edab)


</details>

### MR Checklist
